include/boost/wave/grammars/cpp_has_include_grammar.hpp

```diff
@@ -0,0 +1,215 @@
+    Software License, Version 1.0. (See accompanying file
+    LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+=============================================================================*/
+
+#if !defined(BOOST_CPP_HAS_INCLUDE_GRAMMAR_HPP_F48287B2_DC67_40A8_B4A1_800EFBD67869_INCLUDED)
+#define BOOST_CPP_HAS_INCLUDE_GRAMMAR_HPP_F48287B2_DC67_40A8_B4A1_800EFBD67869_INCLUDED
+
+#include <boost/wave/wave_config.hpp>
+
+#include <boost/assert.hpp>
+#include <boost/spirit/include/classic_core.hpp>
+#include <boost/spirit/include/classic_closure.hpp>
+#include <boost/spirit/include/classic_assign_actor.hpp>
+#include <boost/spirit/include/classic_push_back_actor.hpp>
+
+#include <boost/wave/token_ids.hpp>
+#include <boost/wave/util/pattern_parser.hpp>
+#include <boost/wave/grammars/cpp_has_include_grammar_gen.hpp>
+
+#if !defined(spirit_append_actor)
+#define spirit_append_actor(actor) boost::spirit::classic::push_back_a(actor)
+#endif // !has_include(spirit_append_actor)
+
+// this must occur after all of the includes and before any code appears
+#ifdef BOOST_HAS_ABI_HEADERS
+#include BOOST_ABI_PREFIX
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+namespace boost {
+namespace wave {
+namespace grammars {
+
+///////////////////////////////////////////////////////////////////////////////
+//  define, whether the rule's should generate some debug output
+#define TRACE_CPP_HAS_INCLUDE_GRAMMAR \
+    bool(BOOST_SPIRIT_DEBUG_FLAGS_CPP & BOOST_SPIRIT_DEBUG_FLAGS_HAS_INCLUDE_GRAMMAR) \
+    /**/
+
+template <typename ContainerT>
+struct has_include_grammar :
+    public boost::spirit::classic::grammar<has_include_grammar<ContainerT> >
+{
+    has_include_grammar(ContainerT &tokens_seq_,
+                        bool &is_quoted_filename_,
+                        bool &is_system_)
+    :   tokens_seq(tokens_seq_), is_quoted_filename(is_quoted_filename_),
+        is_system(is_system_), true_(true)
+    {
+        BOOST_SPIRIT_DEBUG_TRACE_GRAMMAR_NAME(*this, "has_include_grammar",
+            TRACE_CPP_HAS_INCLUDE_GRAMMAR);
+        is_quoted_filename = false;
+        is_system = false;
+    }
+
+    template <typename ScannerT>
+    struct definition
+    {
+        typedef boost::spirit::classic::rule<ScannerT> rule_t;
+
+        rule_t has_include_op;
+        rule_t system_include;
+        rule_t nonsystem_include;
+        rule_t computed_include;
+
+        definition(has_include_grammar const & self)
+        {
+            using namespace boost::spirit::classic;
+            using namespace boost::wave;
+            using namespace boost::wave::util;
+
+            has_include_op
+                =   ch_p(T_IDENTIFIER)      // token contains '__has_include'
+                    >>  (
+                        ch_p(T_LEFTPAREN) >>
+                        (system_include | nonsystem_include | computed_include)
+                        )
+                    >> end_p
```

Comment:
What about the right parenthesis? Should this be matched as well?

Reply:
nvm, I found where the right parenthesis is handled. Sorry for the noise.

Reply:
I'm glad you asked :) I initially did, but then moved it [outside the parser](https://github.com/jefftrull/wave/blob/6dc21c848009b8ffea5471b23a21030f9cbe67fa/include/boost/wave/util/cpp_macromap.hpp#L1812) because of a tricky problem with Classic's greedy matching.

I want to accept expressions including nested parentheses while excluding the final right parenthesis. When I tried to do it with the Kleene star it would always grab every right paren, including the one that terminates `__has_include()` - and it does not backtrack when the full expression `ch_p(T_LEFTPAREN) >> ... >> ch_p(T_RIGHTPAREN)` then fails. The explanation for this problem [here](https://www.boost.org/doc/libs/1_73_0/libs/spirit/classic/doc/rationale.html#exhaustive_rd) did not provide any solution I could understand.

Is there a cleaner way to solve this?

Reply:
If you need to accept inner parenthesis a possible solution would be a recursive rule. Something like:
```
r = ch_p(T_LEFTPAREN) >> r >> ch_p(T_RIGHTPAREN) | *char_p;
```